main.cc

```diff
@@ -1065,23 +1054,15 @@ struct play_state : game_state {
 
         // blech. Tool gets used below, then fire projectile gets hit here
         if (pl.fire_projectile) {
-            if (available_projectiles.size()) {
-                auto proj = available_projectiles.back();
-                available_projectiles.pop_back();
-
-                proj->pos = glm::vec3(pl.eye.x, pl.eye.y, pl.eye.z);
-                proj->dir = pl.dir;
-                proj->velocity = 2.f;
-                proj->lifetime = proj->initial_lifetime;
-
-                for (auto i = 0u; i < current_projectiles.size(); ++i) {
-                    if (current_projectiles[i] == nullptr) {
-                        current_projectiles[i] = proj;
-                        break;
-                    }
-                }
+            if (num_projectiles < MAX_PROJECTILES) {
+                auto & proj = projectiles[num_projectiles++];
+                proj.pos = pl.eye;
+                proj.dir = pl.dir;
+                proj.velocity = 2.f;
+                proj.lifetime = proj.initial_lifetime;
+
+                pl.fire_projectile = false;
```

Comment:
moving this inside the if is an improvement as it means if we could not fire this frame we will fire next frame,
we do have to be careful in our sizing of MAX_PROJECTILES in order to try guarantee that in the unlikely moment we cannot fire this frame we are likely to be able to fire next frame or at the least within a few frames.